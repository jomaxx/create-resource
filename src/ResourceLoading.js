import React from "react";

export const Context = React.createContext();

export default class ResourceLoading extends React.Component {
  state = { loading: true };

  componentDidMount() {
    this.setState({ loading: false });
  }

  componentWillUnmount() {}

  startLoading = promise => {
    this.componentWillUnmount();

    let subscribed = true;

    this.componentWillUnmount = () => {
      subscribed = false;
    };

    this.setState(state => {
      if (state.loading) return null;
      return { loading: true };
    });

    promise.then(
      () => subscribed && this.setState({ loading: false }),
      () => subscribed && this.setState({ loading: false })
    );
  };

  render() {
    return (
      <Context.Provider value={this.startLoading}>
        {this.props.children(this.state.loading)}
      </Context.Provider>
    );
  }
}
