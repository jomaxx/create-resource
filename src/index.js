export { default as createResource } from "./createResource";
export { default as ResourceLoading } from "./ResourceLoading";
