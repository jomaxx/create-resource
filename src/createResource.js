import React from "react";
import { Async } from "react-async-await";
import { Context } from "./ResourceLoading";

export default function createResource(load, resolve) {
  const cache = new Map();

  function memoizedLoad(props) {
    const key = resolve(props);
    if (!cache.has(key)) cache.set(key, load(props));
    return cache.get(key);
  }

  class LoadingResource extends React.Component {
    static get displayName() {
      return `Loading(${Resource.displayName || Resource.name || "Resource"})`;
    }

    componentDidMount() {
      this.props.startLoading(this.props.resource.promise);
    }

    componentDidUpdate() {
      this.componentDidMount();
    }

    render() {
      return null;
    }
  }

  function Resource(props) {
    return (
      <Context.Consumer>
        {startLoading => {
          if (!startLoading) {
            throw new Error("missing ResourceLoading ancestor!");
          }

          return (
            <Async
              await={memoizedLoad(props)}
              waiting={promise => ({ status: 0, promise })}
              then={value => ({ status: 1, value })}
              catch={error => ({ status: 2, error })}
            >
              {resource => {
                switch (resource.status) {
                  case 0:
                    return (
                      <LoadingResource
                        startLoading={startLoading}
                        resource={resource}
                      />
                    );
                  case 1:
                    return props.children(resource.value);
                  case 2:
                  default:
                    cache.delete(resolve(props));
                    throw resource.error;
                }
              }}
            </Async>
          );
        }}
      </Context.Consumer>
    );
  }

  return Resource;
}

function noop() {}
