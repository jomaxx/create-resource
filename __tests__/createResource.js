import React from "react";
import { render } from "react-testing-library";
import { createResource, ResourceLoading } from "../src";

function delay(...args) {
  return new Promise(resolve => setTimeout(resolve, ...args));
}

class ErrorBoundary extends React.Component {
  componentWillMount() {
    jest.spyOn(console, "error");
    console.error.mockImplementation(() => {});
  }

  componentDidCatch(error) {
    this.props.onError(error);
  }

  render() {
    return this.props.children;
  }
}

afterEach(() => {
  jest.restoreAllMocks();
});

it("loads resource", done => {
  const Resource = createResource(
    props => delay(100, { id: props.id }),
    props => props.id
  );

  render(
    <ResourceLoading>
      {loading =>
        loading ? null : (
          <Resource id={1}>
            {resource => {
              expect(resource).toEqual({ id: 1 });
              done();
              return null;
            }}
          </Resource>
        )
      }
    </ResourceLoading>
  );
});

it("throws missing ResourceLoading ancestor", done => {
  const Resource = createResource(
    props => ({ id: props.id }),
    props => props.id
  );

  render(
    <ErrorBoundary
      onError={error => {
        expect(error).toMatchSnapshot();
        done();
      }}
    >
      <Resource id={1}>{resource => null}</Resource>
    </ErrorBoundary>
  );
});

it("throws error", done => {
  const Resource = createResource(
    props => delay(100).then(() => Promise.reject(new Error("error"))),
    props => props.id
  );

  render(
    <ErrorBoundary
      onError={error => {
        expect(error).toMatchSnapshot();
        done();
      }}
    >
      <ResourceLoading>
        {() => <Resource id={1}>{resource => null}</Resource>}
      </ResourceLoading>
    </ErrorBoundary>
  );
});

it("can try again after thrown error", async () => {
  const spy = jest.fn();
  const Resource = createResource(spy, props => props.id);

  spy
    .mockImplementationOnce(() => Promise.reject(new Error("error")))
    .mockImplementationOnce(props => ({ id: props.id }));

  await new Promise(resolve => {
    render(
      <ErrorBoundary onError={resolve}>
        <ResourceLoading>
          {() => <Resource id={1}>{resource => null}</Resource>}
        </ResourceLoading>
      </ErrorBoundary>
    );
  });

  await new Promise((resolve, reject) => {
    render(
      <ErrorBoundary onError={reject}>
        <ResourceLoading>
          {() => (
            <Resource id={1}>
              {resource => {
                expect(resource).toEqual({ id: 1 });
                resolve();
                return null;
              }}
            </Resource>
          )}
        </ResourceLoading>
      </ErrorBoundary>
    );
  });
});
