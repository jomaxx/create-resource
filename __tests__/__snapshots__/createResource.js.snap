// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws error 1`] = `[Error: error]`;

exports[`throws missing ResourceLoading ancestor 1`] = `[Error: missing ResourceLoading ancestor!]`;
